LFPy/eegmegcalc.py

```diff
@@ -636,28 +660,34 @@ def calc_phi(self, p_tan):
             in units of (radians) between x-axis vector(s) and projection of
             contact point location vector(s) rxyz into xy-plane.
             z-axis is defined in the direction of rzloc.
-            y-axis is defined in the direction of p (orthogonal to rzloc).
-            x-axis is defined as cross product between p and rzloc (x).
+            y-axis is defined in the direction of p_tan (orthogonal to rzloc).
+            x-axis is defined as cross product between p_tan and rzloc (x).
         """
-        proj_rxyz_rz = (np.dot(self.rxyz, self._rzloc) /
-                        np.sum(self._rzloc ** 2)).reshape(len(self.rxyz),1)\
-                        * self._rzloc.reshape(1,3)
-                        # projection of rxyz onto rzloc
-        rxy = self.rxyz - proj_rxyz_rz  # projection of rxyz into xy-plane
-        x = np.cross(p_tan, self._rzloc)  # vector giving direction of x-axis
-
-        denominator = np.dot(np.linalg.norm(rxy, axis=1).reshape(len(rxy),1),
-                             np.linalg.norm(x, axis=1).reshape(1, len(x)))
-        denominator[np.where(denominator == 0)] = 1e-12
-        cos_phi = np.dot(rxy, x.T) / denominator
-        cos_phi = np.nan_to_num(cos_phi)
-        phi_temp = np.arccos(cos_phi) # nb: phi_temp is in range [0, pi]
-        phi = phi_temp
-        range_test = np.dot(rxy, p_tan.T)  # if range_test < 0, phi > pi
-        for i in range(len(self.r)):
-            for j in range(len(p_tan)):
-                if range_test[i,j] < 0:
-                    phi[i,j] = 2*np.pi - phi_temp[i,j]
+
+        # project rxyz onto z-axis (rzloc)
+        proj_rxyz_rz = np.outer(np.dot(self.rxyz, self._rzloc) / self._rz ** 2, self._rzloc)
+        # find projection of rxyz in xy-plane
+        rxy = self.rxyz - proj_rxyz_rz
+        # define x-axis
+        x = np.cross(p_tan, self._rzloc)
+
+        phi = np.zeros((len(self.rxyz), len(p_tan)))
+        # phi is not defined when theta= 0,pi or |p_tan| = 0
+        mask = np.ones(phi.shape, dtype=bool)
+        mask[np.where((self._theta == 0) |
+                      (self._theta == np.pi))] = np.zeros(len(p_tan))
+        mask[:,np.where(np.abs(np.linalg.norm(p_tan, axis=1)) == 0)] = 0
+        cos_phi = np.zeros(phi.shape)
+        # compute cos_phi using mask to avoid zerodivision
+        cos_phi[mask] = np.dot(rxy, x.T)[mask] / np.outer(np.linalg.norm(rxy,
+                  axis=1), np.linalg.norm(x, axis=1))[mask]
+        # compute phi in [0, pi]
+        phi[mask] = np.arccos(cos_phi[mask])
+
+        # nb: phi in [-pi, pi]. since p_tan defines direction of y-axis,
+        # phi < 0 when rxy*p_tan < 0
+        phi[np.where(np.dot(rxy, p_tan.T) < 0)] *= -1
```

Comment:
`phi[np.dot(rxy, p_tan.T) < 0] *= -1` should do the same as `phi` itself is an array type. In general `np.where` can be avoided altogether.